ttnn/cpp/ttnn/operations/normalization/layernorm/layernorm.hpp

```diff
@@ -0,0 +1,130 @@
+
+struct LayerNorm {
+    static inline const std::array<ttnn::TensorSchema, 4> input_tensor_schemas() {
+        return {
+            ttnn::TensorSchema{
+                2,
+                4,
+                {ttnn::bfloat16, ttnn::bfloat8_b, ttnn::bfloat4_b},
+                {ttnn::TILE_LAYOUT},
+                true,
+                false,
+                false,
+                false},
+            ttnn::TensorSchema{
+                1,
+                4,
+                {ttnn::bfloat16, ttnn::bfloat8_b, ttnn::bfloat4_b},
+                {ttnn::TILE_LAYOUT, ttnn::ROW_MAJOR_LAYOUT},
+                true,
+                false,
+                false,
+                true},
+            ttnn::TensorSchema{
+                1,
+                4,
+                {ttnn::bfloat16, ttnn::bfloat8_b, ttnn::bfloat4_b},
+                {ttnn::TILE_LAYOUT, ttnn::ROW_MAJOR_LAYOUT},
+                true,
+                false,
+                false,
+                true},
+            ttnn::TensorSchema{
+                1,
+                4,
+                {ttnn::bfloat16, ttnn::bfloat8_b, ttnn::bfloat4_b},
+                {ttnn::TILE_LAYOUT, ttnn::ROW_MAJOR_LAYOUT},
+                true,
+                false,
+                false,
+                true}};
+    }
+
+    template <typename... Args>
+    static auto input_tensors_to_validate(
+        const Tensor& input_tensor,
+        float epsilon = 1e-12,
+        const std::optional<const ttnn::Tensor>& weight = std::nullopt,
+        const std::optional<const ttnn::Tensor>& bias = std::nullopt,
+        const std::optional<const ttnn::Tensor>& residual_input_tensor = std::nullopt,
+        Args&&... args) {
+        return std::forward_as_tuple(input_tensor, weight, bias, residual_input_tensor);
+    }
+
+    static inline ttnn::Tensor execute_on_worker_thread(
+        const ttnn::Tensor& input_tensor,
+        float epsilon = 1e-12,
+        const std::optional<const ttnn::Tensor>& weight = std::nullopt,
+        const std::optional<const ttnn::Tensor>& bias = std::nullopt,
+        const std::optional<const ttnn::Tensor>& residual_input_tensor = std::nullopt,
+        const std::optional<MemoryConfig>& memory_config_arg = std::nullopt,
+        const std::optional<const LayerNormProgramConfig>& program_config_arg = std::nullopt) {
+        const LayerNormProgramConfig& program_config = program_config_arg.value_or(LayerNormDefaultProgramConfig{});
+
+        auto memory_config = memory_config_arg.value_or(input_tensor.memory_config());
+        if (residual_input_tensor.has_value()) {
+            return tt::operations::primary::add_layernorm(
+                input_tensor, residual_input_tensor.value(), epsilon, weight, bias, memory_config, program_config);
+        } else {
+            return tt::operations::primary::layernorm(
+                input_tensor, epsilon, weight, bias, memory_config, program_config);
+        }
+    }
+};
+
+struct RMSNorm {
```

Comment:
should ideally be in its own little file

Reply:
will change